crates/walrus-upload-relay/src/tip/check.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! The proxy side of the registration and tipping system.
+//!
+//! It is responsible for validating that a given transaction contains appropriate balance changes
+//! to the associated address.
+
+use std::{
+    num::NonZeroU16,
+    str::FromStr,
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use sui_sdk::{SUI_COIN_TYPE, rpc_types::SuiTransactionBlockResponse};
+use sui_types::TypeTag;
+use tracing::Level;
+use walrus_sdk::core::EncodingType;
+
+use crate::tip::{config::TipConfig, error::TipError};
+
+/// Checks the freshness of the transaction.
+///
+/// Returns an error if the transaction was sent too far in the past.
+#[tracing::instrument(level = Level::DEBUG, skip_all, fields(tx_id=%response.digest))]
+pub(crate) fn check_tx_freshness(
+    response: &SuiTransactionBlockResponse,
+    freshness_threshold: Duration,
+    max_future_threshold: Duration,
+) -> Result<(), TipError> {
+    let Some(tx_timestamp_ms) = response.timestamp_ms else {
+        return Err(TipError::UnexpectedResponse(
+            "the transaction does not have a timestamp; it has not been executed",
+        ));
+    };
+    let tx_timestamp = Duration::from_millis(tx_timestamp_ms);
+    let cur_timestamp = SystemTime::now()
+        .duration_since(UNIX_EPOCH)
+        .expect("clocks are set correctly");
+
+    if let Some(diff) = tx_timestamp.checked_sub(cur_timestamp) {
+        // The timestamp is in the future.
+        tracing::warn!(
+            ?tx_timestamp,
+            ?cur_timestamp,
+            "a transaction with a timestamp in the future was received"
+        );
+        if diff > max_future_threshold {
+            tracing::error!(
+                ?diff,
+                "the transaction is in the future by more than the allowed threshold"
+            );
+            return Err(TipError::UnexpectedResponse(
+                "the transaction is in the future by more than the allowed threshold",
+            ));
+        }
+    };
+
+    if let Some(diff) = cur_timestamp.checked_sub(tx_timestamp) {
+        if diff > freshness_threshold {
+            tracing::debug!(?diff, "the transaction is too old");
+            return Err(TipError::TxTooOld);
+        }
+    }
+
+    Ok(())
+}
+
+/// Checks if the execution results are sufficient to cover the tip.
+pub(crate) fn check_response_tip(
```

Comment:
Question: does the tip have to be in the same PTB that register the blob or not? If I read the doc correctly, the tip can be any standalone transaction, is that correct? If so, if someone pay the service address once, but reuse that transaction as evidence of payment over and over, can this be detected?

Reply:
No, the tip transaction just needs to be built with the correct convention (a pure input in the first input slot containing the hashed auth package. The auth package contains a nonce, the blob digest, and its unencoded length, so it cannot be used to spoof payment for other blobs.

Reply:
Ah, I see. Make sense. The pure input has to be the first input, or any pure input? I guess you will provide a "pay tip transaction" for users to use?

Reply:
In the current design, the pure input has to be the first input and contain the auth package hash but if that becomes cumbersome we could allow the input slot to be varying and then provide it as a parameter to the HTTP endpoint so that it would know where to look in the PTB - though for now this seems unnecessary, I think. 

I'm not sure what you mean by 'provide a "pay tip transaction"' exactly. The functionality to hit this endpoint already exists in the TSSDK where the scaffolding there handles creating the tip transaction as part of the blob registration PTB, but like I mentioned, it's _possible_ (and not technically difficult) to pre-register the blob but then separately decide to tip the walrus upload relay operator and use their service for the actual upload.